### DISCUSSION TITLE
Ownership in Frame::render_widget

### DISCUSSION THREAD

**Discussion by cylewitruk** [ratatui/ratatui#592] [Discussion]
Is there a reason why `widget: W` needs to be owned here? Causing me some headaches when trying to keep different screens in the same field as I can't go from a `Box<dyn Widget>` to an owned/sized `Widget`... (i.e. keeping a currently active interchangeable `Widget` in app state).

https://github.com/ratatui-org/ratatui/blob/c597b87f72bc578cdd65d533506db5bdc49af608/src/terminal.rs#L600

Maybe I'm going about things completely wrong and there's another way to achieve this, but nothing I've found during all of today anyway :sweat:

**Comment by joshka** [comment]
Historical reasons to some extent. See the following for some more details:
- https://github.com/ratatui-org/ratatui/discussions/164
- https://github.com/ratatui-org/ratatui/pull/122
- https://github.com/ratatui-org/ratatui/pull/16

TL;DR: widget was designed for `tui-rs` to not be retained. It was intentionally designed as if it was a parameter object created again on each Frame and so should generally be lightweight. We've thought a bit about changing this to pass by ref instead, but that might break TUI / Ratatui app and widget in the wild. It could be time to revisit this however.

Put in terms of your problem, there can't be an active `Widget` because on each frame the `Widget` is created anew. You'll currently need to store something else for the current state. There are a few options for this that are useful:
- implement your own `Component` trait instead of `Widget` and store your sized components
- if there can be only one current screen, then use an enum with a variant for each state
- your component / enum can either just have a method that draws to the frame (skipping the Widget impl), or have a method that creates a widget to represent the component.

For the method that creates a widget you could even fairly easily make the render part of the component by creating a small intermediary like:
```
struct DynamicWidget<'a> {
    render_fn: Box<dyn FnOnce(Rect, &mut Buffer) + 'a>,
}

impl<'a> DynamicWidget<'a> {
    fn new<F: FnOnce(Rect, &mut Buffer) + 'a>(render_fn: F) -> Self {
        Self {
            render_fn: Box::new(render_fn),
        }
    }
}

impl Widget for DynamicWidget<'_> {
    fn render(self, area: Rect, buf: &mut Buffer) {
        (self.render_fn)(area, buf);
    }
}

struct SomeComponent {
    greeting: String,
}

impl SomeComponent {
    fn to_widget<'a>(&'a self) -> DynamicWidget {
        DynamicWidget::new(|area, buf| self.render(area, buf))
    }

    fn render(&self, area: Rect, buf: &mut Buffer) {
        Paragraph::new(self.greeting.as_str()).render(area, buf);
    }
}
```
It's also worth taking a look at https://ratatui.rs/concepts/application-patterns/index.html for some ideas about how to handle this as well as looking at how some larger examples structure their code (I was particularly inspired by [EDMA](https://github.com/nomadiz/edma) when building [toot-rs](https://github.com/toot-rs/toot-rs) and [yazi](https://github.com/sxyazi/yazi) took some of those ideas a bit further. Working out how to handle multiple views was a difficult thing for me to understand when I first started playing with Ratatui, and there's more examples of this out there (see also @kdheepak's [ratatui async template](https://github.com/ratatui-org/ratatui-async-template) as one option).

P.S. I'm going to convert this to a discussion rather than an issue.

**Reply by cylewitruk** [reply]
Thanks a lot for the detailed response! It looks like the `component` pattern is pretty much what I'm trying to achieve - not sure why I didn't stumble across those pattern examples earlier... probably because I thought it was more of a me vs. the borrow checker problem when in actuality I should have been spending more time in the Ratatui book :)

**Reply by cylewitruk** [reply]
@joshka a followup question - I switched to using the component pattern using the async template example (and it works great now, btw!) - but my global fg/bg styling has stopped working... Before I was using:
```rust
let frame = tui.get_frame().size();
tui.current_buffer_mut().set_style(frame, self.theme.main);
```
which worked, but now when I try this (in the async example `tui` is deref to `terminal`), it doesn't seem to work. I've tried placing it in multiple different locations but haven't gotten it to do its thing. Is there something I should be thinking of when it comes to this kind of model and/or when using async?

**Reply by joshka** [reply]
I haven't seen anyone set a style on the entire buffer before, but it that kinda makes sense. I'm not certain how this should render without actually writing to every cell.

Can you expand on "doesn't work"? Errors? Compilation problems? Does nothing? Can you link your code for this? Perhaps a push a couple of branches showing the attempts?

P.S. in the latest release 0.24.0 we expose the buffer directly on the `Frame`, which might be what you need. See https://docs.rs/ratatui/latest/ratatui/terminal/struct.Frame.html#method.buffer_mut, which means you probably can do that without having to call other methods on terminal.

**Reply by kdheepak** [reply]
Does this work?

```rust
let frame = tui.0.get_frame().size();
tui.0.current_buffer_mut().set_style(frame, self.theme.main);
```

My guess is that the issue is because you are trying to mutably borrow buffer and the borrow checker is preventing it for some reason. Like @joshka mentioned, if you have an example you want to share we can help suggest alternatives. And as @joshka said, the `Frame` exposes the buffer directly now. You can use in the `draw` method of a component, which might help you solve your issue.

**Comment by cylewitruk** [comment]
Aha, I had borrowed this method from `rrtop` while using that project as an example and just took for granted that it was some sort of standard way for rendering a default style.

Thanks to your comments, I went back and re-visited the placement of my snippet and realized that it wasn't within the `draw`-loop but rather before it. The new `Frame::buffer_mut()` method definitely helped because I had problems with the borrow checker in 0.23 getting the buffer from within `Terminal::draw()` to set the style.

Here's a snippet of the working code now, which is based on the async-template example :)
```rust
Action::Render => {
   tui.draw(|f| {
      // Set default theme for the frame.
      let frame_size = f.size();
      f.buffer_mut().set_style(frame_size, self.theme.main);

      for component in self.components.iter_mut() {
          let r = component.borrow_mut().draw(f, f.size(), self.theme);
          if let Err(e) = r {
              action_tx.send(Action::Error(format!("Failed to draw: {:?}", e))).unwrap();
          }
      }
   })?;
},
  ```
Thank you for the help!

**Comment by cylewitruk** [comment]
What do you guys usually do with discussions? Should we close it?

**Reply by joshka** [reply]
We don't really have a standard practice - seems ok to close though.
